### ISSUE TITLE
Widgets display changes when moving app to another screen

### ISSUE THREAD

**Issue by cjw85** [kivy/kivy#6785]
**Software Versions**
* Python: 3.6
* OS: MacOS 10.15.3
* Kivy: v1.11.1
* Kivy installation method: pip

**Describe the bug**
When dragging an app from an external display (1920 x 1200) to the internal display of my Macbook (2560x1600), widgets in the app shrink their display size relative to the app window. However their clickable location appears to be where one would expect.

**Expected behavior**
That the widgets scale appropriately to the app window.

**To Reproduce**
Using the example app defined here:
https://kivy.org/doc/stable/guide/lang.html#designing-with-the-kivy-language
start the app on one display and drag it to another of a different resolution.

**Code and Logs and screenshots**
On initial display:
![image](https://user-images.githubusercontent.com/3592009/76786330-f4f6b200-67ae-11ea-9aff-c25728150542.png)
After dragging to second display:
![image](https://user-images.githubusercontent.com/3592009/76786368-0475fb00-67af-11ea-94a7-3d4c238199f7.png)

**Additional context**
No additional context

**Comment by Cheaterman** [comment]
Is your app starting as maximized? Can you create a on_size handler in your App and add a print or logger call inside it?

Did you try to unmaximize-remaximize the window after dragging it to the second display?

**Comment by cjw85** [comment]
Thanks for the feedback. I'll have a look at on_size, if its being fired

The app isn't starting maximised, and I didn't unmaximise-remaximise.

**Comment by Cheaterman** [comment]
Please give the unmax/remax idea an attempt. Other than that, keep us in touch about the on_size event. Thanks!

EDIT: I just realized something... If it's not the size, it could have to do with the dpi of your screens. Your internal display is probably a retina, and if you start the app on the other display, it is "configured" with whatever the pixel density of that screen would be (default 72dpi IIRC?). So when you change screens, the app looks super small.

**Comment by Cheaterman** [comment]
Good or bad news - this seems to be a duplicate of https://github.com/kivy/kivy/issues/6708 .

**Comment by pushfoo** [comment]
If i can find a spare hdmi cable, I'll try to replicate this. It might be a few days as going out to fetch one isn't really an option at the moment. @Cheaterman you may be right that this is another result of the issue i reported.
I've seen some applications stutter when moving between low dpi external screens and the internal retina display. That implies there are some sort of methods for detecting scale changes that they use, but kivy may still using the same texture / surface without rescaling.

**Comment by pushfoo** [comment]
I might have part of a solution for this ticket, but I don't know how or where to implement it.

### tl;dr 
there is an [event on OS X that tells an application that it may need to redraw the window because display density or color space changed.](https://developer.apple.com/documentation/appkit/nswindowdelegate/1419517-windowdidchangebackingproperties?language=objc) Adding a handler for this may allow the display to be redrawn when moving between different dpi contexts.

This ticket may depend on #6708 rather than being a duplicate. #6708 requires fixing a response to an already existing event, but this ticket may require adding a new event listener in c / objective c. 

### details

I looked for other projects that encountered similar issues. Godot seems to have not only a similar issue ( godotengine/godot#5688 ), but a nearly identical display scaling glitch _[(Screenshot source)](https://github.com/godotengine/godot/issues/5688#issuecomment-459108454)_ that seems to result from [opengl screen space starting at the bottom left](https://learnopengl.com/Getting-started/Coordinate-Systems):

![https://github.com/godotengine/godot/issues/5688#issuecomment-459108454](https://user-images.githubusercontent.com/9257572/52012522-a04f8500-248f-11e9-8e0b-3a31e5c9a1d0.png)


On that same issue, @bruvzg [mentions an objective C event that notifies an application that window resolution changed ](https://github.com/godotengine/godot/issues/5688#issuecomment-318647380):

> [windowDidChangeBackingProperties](https://developer.apple.com/documentation/appkit/nswindowdelegate/1419517-windowdidchangebackingproperties?language=objc) may help

That event seems to be the perfect for refreshing the display when a window is moved between screens of different resolutions:
> The notification NSWindowDidChangeBackingPropertiesNotification is posted in macOS 10.7.3 or later when a window’s backing scale factor or its color space changes.

I tried searching for for nswindowdelegate in the kivy and plyer repos, but saw zero results so it may need to be added somewhere in the C / objective C layers. I'm not sure which repo is the right place for this is, and would welcome advice.

**Comment by Cheaterman** [comment]
Thanks a lot for your research! I'm sure it will help tremendously! The thing is, I suppose we're using SDL2 window provider on OSX just like the rest, so the real thing to do here is find out if, where and how they expose this event or information. :-)

**Comment by pushfoo** [comment]
I'm talking to some people about that right now and one of them might be able to give me some guidance on that tomorrow.  From what they said so far, it sounds like you're on the right track  with the idea that SDL2 should handle that OS X context switch event somehow.

**Comment by pushfoo** [comment]
Additional notes on what to read if you're not familiar:

I've also been told to try looking at:
 - [kivy/core/window/window_sdl2.py](https://github.com/kivy/kivy/blob/master/kivy/core/window/window_sdl2.py)
- [kivy/core/window/_window_sdl2.pyx](https://github.com/kivy/kivy/blob/master/kivy/core/window/_window_sdl2.pyx)

I think these tickets may also be related:
 - #3311
- #6144
- #6446

The comment threads on the first two are informative and may be good placers to start understanding the internal structure a bit better.

**Comment by cjw85** [comment]
Those tickets do look to be the same and related symptoms. Given that the first one is from 5 years ago I'm going to guess that I shouldn't hold my breath on a fix here.

**Comment by pushfoo** [comment]
I finally managed to borrow an HDMI cable. I can't get the script to launch the window on my external monitor, but i can confirm that moving a window from the internal monitor to the external one causes font size to suddenly increase relative to the screen area. I would appreciate it if someone can help me force the window to launch on a specific display.

- Python : 3.7.7
- Kivy : 1.11.1
- OS : OS X 10.12.6
- Kivy Installation method : pip

The event I mentioned earlier, windowDidChangeBackingProperties, [turns out to exist in pyobjus already](https://github.com/kivy/pyobjus/blob/672ed6191df577545f11f838e35f7095330149f0/pyobjus/protocols.py#L1353). However, kivy also may need to get updated to have a more consistent idea of dpi. I think there was a pull request that partially fixed both of these (https://github.com/kivy/kivy/pull/6289), but it fell behind and was forgotten.

**Comment by jeffreyrdcs** [comment]
New Kivy user here. I encountered the same issue and stumbled upon this ticket. In my case, the app font size will increase relative to the screen size when I move the window from my internal retina display to an external one. 
I end up trying to use on_resize to let kivy detect which monitor the screen is in and rescale the text according to that. Would be nice to know if there is a better fix. Here are some screenshots of the issue and my code snippet.  The last button is the one that resizes accordingly with monitor change, it also resizes itself when you resize the screen.

Screen in internal display:
<img width="562" alt="Kivy_retina_display" src="https://user-images.githubusercontent.com/10601455/107804993-9d651b80-6d5c-11eb-86dd-021573a0309b.png">
Screen in external display:
![Kivy_external_display](https://user-images.githubusercontent.com/10601455/107805015-a0f8a280-6d5c-11eb-9a31-fe3999702f74.png)


kivy_display_text_test.py:

```
from kivy.app import App

from kivy.uix.boxlayout import BoxLayout
from kivy.uix.label import Label
from kivy.uix.button import Button
from kivy.properties import NumericProperty
from kivy.metrics import dp, sp
from kivy.core.window import Window

wininit = (450,650)
Window.size = wininit

class LoginScreen(BoxLayout):

	test_font_size = NumericProperty(40)
	base_font_size = NumericProperty(40)

	def __init__(self, *args):
		super(LoginScreen, self).__init__()
		self.window_size_safe = [0, 0]
		self.window_pixel_dim_safe = [0, 0]
		self.display_num = 0
		self.window_width_min = wininit[0]
		self.window_height_min = wininit[1]
		Window.bind(on_resize = self.resolution_check_update)

	def resolution_check_update(self, instance, x, y):
		#Properties that I found:
		#0. I notice that in retina display the actual height and width in pixel is actually a multiple of the window.size (?)
		#1. Moving the window to an external display doesn't change get_system_size().
		#2. Rescaling the window changes system.size instantly.
		#3. on_resize gets triggered on both resizing the screen and switching displays.
		#4. During initization, self.height and self.width gives 1600, 1200, but x, y give correct sizes (Don't understand why).
		#5. The self.height and self.width values lag. It only gets set AFTER we finish resizing the window and confirm the size.
		#6. Looking through the code I discovered that there is this x,y that return from on_resize, which seems to be the instanteous size that get updated faster than height,width
		# get_system_size = Window._get_system_size()
		# selfheightwidth = [int(self.width), int(self.height)]
		# print(f'Get_System_size: {get_system_size},  SelfHW: {selfheightwidth},  System_size: {Window.system_size}, X:{x}, Y:{y}')

		# Initialize the values when it first launches
		if self.window_size_safe == [0, 0]:
			self.window_size_safe = Window._get_system_size()
			self.window_pixel_dim_safe = [x, y]
			print(f'Saved the initiated size_safe: {self.window_size_safe}')
			return 

		# Check whether the resize is caused by switching monitors or 
		if Window._get_system_size() == self.window_size_safe:
			print(f'Window moved to another display: {Window.size}')
			window_current_pixel_dim = [x, y]

			# Check if the window pixel dimensions suddenly changes (in my retina display it is 2x less when move to external) 
			if self.window_pixel_dim_safe == multiply_list(window_current_pixel_dim, 2):
				print('Resizing font for External Display')
				self.display_num = 1

			# Check if the window pixel dimensions suddenly changes (in my retina display it is 2x more when move to retina) 				
			elif self.window_pixel_dim_safe == multiply_list(window_current_pixel_dim, 0.5):
				print('Resizing font for Retina Display')
				self.display_num = 0

			# Update the font_size, x,y change should be identical
			scale_fac = window_current_pixel_dim[0]/self.window_pixel_dim_safe[0]
			self.test_font_size = self.test_font_size * scale_fac

			# Save it to base_font for font resizing when window size changes
			self.base_font_size = self.test_font_size

			# Saved the current window pixel dimensions
			self.window_pixel_dim_safe = window_current_pixel_dim

		else:
			window_current_pixel_dim = [x, y]			
			print(f'Window being resized: {Window.size}')
			# Update the font size (Enlarge font size after window resize)
			# Idea from here: https://www.reddit.com/r/kivy/comments/86okok/update_font_sizes_in_a_layoutwidget_after_window/
			if self.display_num == 0:
				scale_fac_display = 2
			elif self.display_num == 1:
				scale_fac_display = 1
			scale_fac = min(window_current_pixel_dim[0]/(scale_fac_display * self.window_width_min), window_current_pixel_dim[1]/(scale_fac_display * self.window_height_min))
			self.test_font_size = self.base_font_size * scale_fac
			print(f'Scale fac {scale_fac}')
			self.window_pixel_dim_safe = window_current_pixel_dim
			self.window_size_safe = Window._get_system_size()
			

def multiply_list(in_list, value):
	return [element * value for element in in_list]


class DisplayTextTestApp(App):
	Window.minimum_width = wininit[0]
	Window.minimum_height = wininit[1]
	def build(self):
		return LoginScreen()	


if __name__ == "__main__":
	DisplayTextTestApp().run()




```

displaytexttest.kv

```
#:kivy 2.0.0

<LoginScreen>
	canvas.before:
		Color: 
			rgba: [1., 1., 1., 1]
		Rectangle:
			pos: self.pos
			size: self.size

	orientation: "vertical"
	spacing: 10
	padding: [self.width * 0.05, self.height * 0.05]      # Need to be a percentage of height/width

	Label:
		text: 'TESTING LABEL TEXT'
		size_hint: (1, 0.3)
		font_size: 35
		color: [0, 0, 0, 1]

	Button:
		text: 'This is 30pt'
		size_hint: 0.5, 0.05
		font_size: 30
		pos_hint: {'center_x': 0.5}
		background_normal: ''
		background_color: [0.906, 0.906, 0.906, 1]
		color: [0, 0, 0, 1]

	Button:
		text: 'This is 30dp'
		size_hint: 0.5, 0.05
		font_size: dp(30)
		pos_hint: {'center_x': 0.5}
		background_normal: ''
		background_color: [0.906, 0.906, 0.906, 1]
		color: [0, 0, 0, 1]

	Button:
		text: 'This is 30sp'
		size_hint: 0.5, 0.05
		font_size: sp(30)
		pos_hint: {'center_x': 0.5}
		background_normal: ''
		background_color: [0.906, 0.906, 0.906, 1]
		color: [0, 0, 0, 1]

	Button:
		text: 'This one resizes!'
		size_hint: (0.5, 0.05)
		font_size: root.test_font_size
		pos_hint: {'center_x': 0.5}
		background_normal: ''
		background_color: [0.906, 0.906, 0.906, 1]
		color: [0, 0, 0, 1]
```

**Comment by matham** [comment]
This should be fixed in master with https://github.com/kivy/kivy/pull/7293. Please take a look.

**Comment by jeffreyrdcs** [comment]
Thanks, I just took a look at #7293. That's a lot of work.  If you don't mind I do have one question.
From what I understand #7293 implement an update on all the Metrics density related NumericProperty instances such as font_scale etc, so once Metrics.dpi / density changes all the relevant parameters "x dp" get recomputed automatically.  But if I understand correctly, Metrics.density is not going to update itself real-time when we switch display (issues in this ticket)?
Does that mean we still have to include the detection part where we find out which display the screen is on, and then change the Metrics density accordingly?

Using the snippet I post above as an example, should I update it to something like this, which should work, or is there a cleaner way?: 
```
		# Check whether the resize is caused by switching monitors or resizing
		if Window._get_system_size() == self.window_size_safe:
			print(f'Window moved to another display: {Window.size}')
			# Check if the window pixel dimensions suddenly changes (in my retina display it is 2x less when move to external) 
			if self.window_pixel_dim_safe == multiply_list(window_current_pixel_dim, 2):
				print('Resizing font for External Display')
				# Update the Metric.density or dpi (after #7293)
				Metrics.density = 1
				self.display_num = 1

			# Check if the window pixel dimensions suddenly changes (in my retina display it is 2x more when move to retina) 				
			elif self.window_pixel_dim_safe == multiply_list(window_current_pixel_dim, 0.5):
				print('Resizing font for Retina Display')
				# Update the Metric dpi (after #7293)
				Metrics.density = 2
				self.display_num = 0
		else:
                       print(f'Window being resized: {Window.size}')
```

**Comment by matham** [comment]
The density does update dynamically with DPI as the screen is changed (on osx, on windows you need to make the ctypes call for it to happen) causing the layout to be retriggered and any other DPI/density sensitive properties to be re-computed.

**Comment by Fretless14** [comment]
Hi,

I'm still having this exact issue, but only on Mac (or a high density retina display, idk what the issue is).

The image is halved, however all of the buttons and what not are in the place they are supposed to, so you can control the app by guessing where the buttons are without seeing them.

**Works on my Windows machine with a 3840x2160 display to a 1080p display** (and vice versa) but not for the Mac/retina display to the 1080p monitor (and vice versa). Going from the Mac display to the 1080p enlarges the image+buttons, and the opposite halves the whole image (but buttons are in correct place).

The buttons might be working in the correct spot because I'm using KivyMD (just a thought)

Python 3.9
Kivy==2.3.0
